From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TheDBP <TheDBP@users.noreply.github.com>
Date: Mon, 21 Sep 2026 10:24:14 -0400
Subject: [PATCH] Revert "init: remove session keyring workaround for old
 kernels"

This reverts commit 5d7c35ce2c6f8b7a4bfcf8c4ea35a4a4a1f55fe8.

Kernels without FS_IOC_ADD_ENCRYPTION_KEY (fscrypt before 5.4, e.g.
msm-4.9) take v1 keys through the session keyring, which vold expects
init to have created before installkey. Goes with the vold revert.
Conflicts: Android.bp only (libhidl-gen-utils had been dropped since;
re-add libkeyutils alone).
---
 init/Android.bp                  |  1 +
 init/builtins.cpp                |  3 +++
 init/fscrypt_init_extensions.cpp | 16 ++++++++++++++++
 init/fscrypt_init_extensions.h   |  1 +
 init/fuzzer/Android.bp           |  1 +
 init/init.cpp                    |  6 ++++++
 6 files changed, 28 insertions(+)

diff --git a/init/Android.bp b/init/Android.bp
index 703542f34..5df203630 100644
--- a/init/Android.bp
+++ b/init/Android.bp
@@ -206,6 +206,7 @@ libinit_cc_defaults {
         "libext4_utils",
         "libfs_mgr",
         "libgsi",
+        "libkeyutils",
         "liblog",
         "liblogwrap",
         "liblp",
diff --git a/init/builtins.cpp b/init/builtins.cpp
index 2e4a3b4dc..8f34ce3ca 100644
--- a/init/builtins.cpp
+++ b/init/builtins.cpp
@@ -594,6 +594,9 @@ static Result<void> queue_fs_event(int code) {
     } else if (code == FS_MGR_MNTALL_DEV_FILE_ENCRYPTED ||
                code == FS_MGR_MNTALL_DEV_IS_METADATA_ENCRYPTED ||
                code == FS_MGR_MNTALL_DEV_NEEDS_METADATA_ENCRYPTION) {
+        if (!FscryptInstallKeyring()) {
+            return Error() << "FscryptInstallKeyring() failed";
+        }
         SetProperty("ro.crypto.state", "encrypted");
 
         // Although encrypted, vold has already set the device up, so we do not need to
diff --git a/init/fscrypt_init_extensions.cpp b/init/fscrypt_init_extensions.cpp
index 6a561e54c..fbd818957 100644
--- a/init/fscrypt_init_extensions.cpp
+++ b/init/fscrypt_init_extensions.cpp
@@ -34,12 +34,28 @@
 #include <cutils/properties.h>
 #include <cutils/sockets.h>
 #include <fscrypt/fscrypt.h>
+#include <keyutils.h>
 #include <logwrap/logwrap.h>
 
 #define TAG "fscrypt"
 
 using namespace android::fscrypt;
 
+bool FscryptInstallKeyring() {
+    if (keyctl_search(KEY_SPEC_SESSION_KEYRING, "keyring", "fscrypt", 0) != -1) {
+        LOG(INFO) << "Keyring is already created";
+        return true;
+    }
+    key_serial_t device_keyring = add_key("keyring", "fscrypt", 0, 0, KEY_SPEC_SESSION_KEYRING);
+
+    if (device_keyring == -1) {
+        PLOG(ERROR) << "Failed to create keyring";
+        return false;
+    }
+    LOG(INFO) << "Keyring created with id " << device_keyring << " in process " << getpid();
+    return true;
+}
+
 // TODO(b/139378601): use a single central implementation of this.
 static void delete_dir_contents(const std::string& dir) {
     char* const paths[2] = {const_cast<char*>(dir.c_str()), nullptr};
diff --git a/init/fscrypt_init_extensions.h b/init/fscrypt_init_extensions.h
index 5e0269a3b..d357bb2fd 100644
--- a/init/fscrypt_init_extensions.h
+++ b/init/fscrypt_init_extensions.h
@@ -25,5 +25,6 @@ enum class FscryptAction {
     kDeleteIfNecessary,
 };
 
+bool FscryptInstallKeyring();
 bool FscryptSetDirectoryPolicy(const std::string& ref_basename, FscryptAction action,
                                const std::string& dir);
diff --git a/init/fuzzer/Android.bp b/init/fuzzer/Android.bp
index 355c8d8ba..882b1f575 100644
--- a/init/fuzzer/Android.bp
+++ b/init/fuzzer/Android.bp
@@ -30,6 +30,7 @@ cc_defaults {
     shared_libs: [
         "libbase",
         "libfs_mgr",
+        "libkeyutils",
         "liblog",
         "libprocessgroup",
         "libselinux",
diff --git a/init/init.cpp b/init/init.cpp
index aec2688fc..34174e503 100644
--- a/init/init.cpp
+++ b/init/init.cpp
@@ -54,6 +54,7 @@
 #include <com_android_apex_flags.h>
 #include <fs_avb/fs_avb.h>
 #include <fs_mgr_vendor_overlay.h>
+#include <keyutils.h>
 #include <libavb/libavb.h>
 #include <libdm/loop_control.h>
 #include <libgsi/libgsi.h>
@@ -1106,6 +1107,11 @@ int SecondStageMain(int argc, char** argv) {
                    << " to /proc/1/oom_score_adj: " << result.error();
     }
 
+    // Set up a session keyring that all processes will have access to. It
+    // will hold things like FBE encryption keys. No process should override
+    // its session keyring.
+    keyctl_get_keyring_ID(KEY_SPEC_SESSION_KEYRING, 1);
+
     // Indicate that booting is in progress to background fw loaders, etc.
     close(open("/dev/.booting", O_WRONLY | O_CREAT | O_CLOEXEC, 0000));
 
